AndroidHomeApplication/app/src/main/java/com/example/androidhomeapplication/reduxPagination/Store.kt

```diff
@@ -0,0 +1,103 @@
+package com.example.androidhomeapplication.reduxPagination
+
+class Store<T> {
+    private var state: State = State.Empty
+
+    var render: (State) -> Unit = {}
+    var sideEffects: (SideEffect) -> Unit = {}
+
+    fun proceed(action: Action) {
+        val newState = reducer<T>(action, state) { sideEffect ->
+            sideEffects(sideEffect) // вынести на другой поток
+        }
+        if (newState != state) {
+            state = newState
+            render(state)
+        }
+    }
+
+    private fun <T> reducer(
+        action: Action,
+        state: State,
+        sideEffectListener: (SideEffect) -> Unit
+    ): State = when (action) {
+        is Action.Restart -> {
+            sideEffectListener(SideEffect.LoadPage(1))
+            State.EmptyProgress
+        }
+        is Action.Refresh -> {
+            when (state) {
+                is State.Empty -> {
+                    sideEffectListener(SideEffect.LoadPage(1))
+                    State.EmptyProgress
+                }
+                is State.Data<*> -> {
+                    sideEffectListener(SideEffect.LoadPage(1))
+                    State.Refresh(state.page, state.data as List<T>)
```

Comment:
to avoid all such unnecerry casts
just make Action<T> and State<T>
cases with objects make with
: Action<Nothing>